File: eclipse/src/utilities/adt/moe/Stack.java
package utilities.adt.moe;

/**
 * @author MOE
 * @param <T> The generic type which the this stack stores.
 */
public class Stack<T> {
    
	/** The field which points to the first item (node) in the stack. */
	private Node<T> head;
    
    /** C'tor which creates an empty stack and returns it. */
    public Stack() {
        this.head = null;
    }
    
    /** Creates a new item (node) with the given object and insert it
     	as the stack's new first item.  */
    public void push(T x) {
        Node<T> temp = new Node<T>(x);
        temp.setNext(head);
        head = temp;
    }
    
    /** Remove the stack's first item (node) and returns it. */
    public T pop() {
    	T x = head.getValue();
        head = head.getNext();
        return x;
    }

    /** Returns the first item - without removing it from the stack. */
    public T top() {
        return head.getValue();
    }

    /** Returns `true` if the stack is empty, otherwise - `false`. */
    public boolean isEmpty() {
        return head == null;
    }

    /** Returns a string which represents the stack currently. */
    public String toString() {
        String s = "[";
        Node<T> p = this.head;
        while (p != null)
        {
            s = s + p.getValue().toString();
            if (p.getNext() != null)
                s = s + ",";
            p = p.getNext();
        }
        s = s + "]";
        return s;
    }
}

File: VS/utils/Stack/Stack.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace MOE {
	public class Stack<T> {

		// תכונת המחלקה
		private Node<T> first;

		// פעולה בונה היוצרת ומחזירה מחסנית ריקה
		public Stack() {
			this.first = null;
		}

		// פעולה המחזירה `אמת` אם המחסנית ריקה ו-`שקר` אחרת
		public bool IsEmpty() {
			return this.first == null;
		}

		// דחיפה למחסנית - פעולה הדוחפת איבר לראש המחסנית
		public void Push(T x) {
			this.first = new Node<T>(x, this.first);
		}

		// שלוף ממחסנית - פעולה המוציאה ומחזירה את האיבר שבראש המחסנית
		// הנחה: המחסנית אינה ריקה
		public T Pop() {
			T x = this.first.GetValue();
			this.first = this.first.GetNext();
			return x;
		}

		// הצץ למחסנית - פעולה המחזירה את האיבר שבראש המחסנית מבלי להוציאו
		// הנחה: המחסנית לא ריקה
		public T Top() {
			return this.first.GetValue();
		}

		// פעולה המחזירה מחרוזת המתארת את מצב המחסנית באופן הבא
		// [x1, x2, ..., xn]
		// x1 כאשר בראש המחסנית נמצא האיבר
		// xn ואילו בסוף המחסנית נמצא האיבר
		public override string ToString() {
			string str = "[";
			Node<T> pos = this.first;

			while (pos != null) {
				str += pos.GetValue().ToString();
				if (pos.HasNext())
					str += ", ";

				pos = pos.GetNext();
			}

			str += "]";

			return str;
		}
	}
}
